### DISCUSSION TITLE
How to get the largest value less than a specified value from a list of integers

### DISCUSSION THREAD

**Discussion by AbhaMoitra** [SemanticApplicationDesignLanguage/sadl#776] [Q&A] [answered]
I am trying to work with lists and while I can take the max of a list of integers, I need to get the max of the list of integers where the integers are less than a specified value. I have been unable to do this, any ideas / suggestions. Here is a simple project of what I have.

uri "http://sadl.org/TestList.sadl".
 
TestClass is a class
 	described by snapshot with values of type PropSnapshot
 	described by ind with a single value of type int.
PropSnapshot is a class  
	described by prop1 with a single value of type int
 	described by prop2 with a single value of type int. 	

// For instance m1, the prop1 values are [3, 5, 8, 10].
// What I want to do is if I am supplied with a value x (via property ind) then I want to find 
// the prop1 value that is largest value less than x.
// So if x is 9 I want to print 8.
// In Rule Test1, I am trying to do it via a list. I can get the max of prop1 values 
// but I am not able to add in the clause "val1 < ind1" without getting an error.
Rule Test1
if mat1 is a TestClass
   and ind of mat1 is ind1
   and list1 is list(mat1, snapshot, ss, ss, prop1, val1) 
   // the following is my attempt but this gives error
//   and list1 is list(mat1, snapshot, ss, ss, prop1, val1, val1, lessThan, ind1)   
   and out is max(list1)
then print("in Test1, ind1 and out are", ind1, out).

// Trying to do it without using list, but again no able to do so
Rule Test2
if mat1 is a TestClass
   and ind of mat1 is ind1
   and snapshot of mat1 is ss1
   and prop1 of ss1 is candidateVal
   and candidateVal < ind1
   and snapshot of mat1 is ss2
   and prop1 of ss2 is anotherVal
   and anotherVal < ind1
   // want to say that there is no value of anotherVal that is greater than candidateVal
then print("in Test2, ind1 and out are", ind1, candidateVal).

m1 is a TestClass, 
  has snapshot (a PropSnapshot 
  	               with prop1 3,
  	               with prop2 9)
  has snapshot (a PropSnapshot 
  	               with prop1 5,
  	               with prop2 10)
  has snapshot (a PropSnapshot 
  	               with prop1 8,
  	               with prop2 15)
  has snapshot (a PropSnapshot 
  	               with prop1 10,
  	               with prop2 20)
  has ind 9.

Ask: "select * where {?x ?y ?z} limit 1".

**Comment by crapo** [comment, marked as answer]
@AbhaMoitra , I think that you must either separate the comparison operation from the list operation or else create a custom built-in function that combines the two. Following the former approach, consider this model:

```
 uri "http://sadl.org/TestList2.sadl" alias testlist2.
 
 TestClass is a class
described by snapshot with values of type PropSnapshot
described by ind with a single value of type int.
PropSnapshot is a class
described by prop1 with a single value of type int
described by prop2 with a single value of type int.

valueLessThanThreshold describes TestClass with values of type int.
maxLessThanThreshold describes TestClass with a single value of type int.

Rule R1:
	if matl is a TestClass
		and ind1 is ind of matl
		and pss1 is snapshot of matl
		and val1 is prop1 of pss1
		and val1 < ind1
	then
		print("R1: adding ", val1, " valuesLessThanThreshold of TestClass ", matl) and
		matl has valueLessThanThreshold val1.
		
Rule R2:
	if matl is a TestClass
		and ind1 is ind of matl
		and vltt is valueLessThanThreshold of matl
		and list1 is list(matl, valueLessThanThreshold)
		and print("R2: values less than threshold are ", listToString(list1))
		and out is max(list1)
	then maxLessThanThreshold of matl is out and
		print("R2: TestClass ", matl, " has value ", out, " less than ", ind1).
		
m1 is a TestClass,
has snapshot (a PropSnapshot
with prop1 3,
with prop2 9)
has snapshot (a PropSnapshot
with prop1 5,
with prop2 10)
has snapshot (a PropSnapshot
with prop1 8,
with prop2 15)
has snapshot (a PropSnapshot
with prop1 10,
with prop2 20)
has ind 9.

Ask: "select ?tc ?mltt where {?tc <rdf:type> <TestClass> . 
		?tc <maxLessThanThreshold> ?mltt}".
```

When executed, the following is output to the console window:
```
Inference of '/home/camfe/andy/ws/sadl/runtime-EclipseApplication2/GH-776/TestList2.sadl' requested.
print: R1: adding  8  to valueLessThanThreshold of TestClass  <http://sadl.org/TestList2.sadl#m1> 
print: R2: values less than threshold are  {8} 
print: R2: TestClass  <http://sadl.org/TestList2.sadl#m1>  has maxLessThanThreshold value  8  which is less than  9 
print: R1: adding  5  to valueLessThanThreshold of TestClass  <http://sadl.org/TestList2.sadl#m1> 
print: R2: values less than threshold are  {5, 8} 
print: R2: TestClass  <http://sadl.org/TestList2.sadl#m1>  has maxLessThanThreshold value  8  which is less than  9 
print: R1: adding  3  to valueLessThanThreshold of TestClass  <http://sadl.org/TestList2.sadl#m1> 
print: R2: values less than threshold are  {3, 5, 8} 
print: R2: TestClass  <http://sadl.org/TestList2.sadl#m1>  has maxLessThanThreshold value  8  which is less than  9 
Reasoner family: Jena-Based (com.ge.research.sadl.jena.reasoner.JenaReasonerPlugin), version date 12 April 2021.
Query: select ?tc ?mltt where {?tc <rdf:type> <TestClass> . 
		?tc <maxLessThanThreshold> ?mltt}
     "tc","mltt"
     "m1","8"
```

I'm concerned that the print function outputs are repeated with incomplete lists of values, but then there is only one value assigned for maxLessThanThreshold, so that leaves me a little confused. The 3rd line of R2 is meant to tell the Rete Engine that R1 should be fired to completion before R2 is fired, which appears to be happening per the final outcome but placed in question by the print function output.
